mining: add getMemoryLoad() and track template non-mempool memory footprint

Sjors opened the pull request:
Implements a way to track the memory footprint of all non-mempool transactions that are still being referenced by block templates, see discussion in  #33899. It does not impose a limit.

IPC clients can query this footprint (total, across all clients) using the `getMemoryLoad()` IPC method. Its client-side usage is demonstrated here:
- https://github.com/stratum-mining/sv2-tp/pull/63

Additionally, the functional test in `interface_ipc.py` is expanded to demonstrate how template memory management works: templates are not released until the client drops references to them, or calls the template destroy method, or disconnects. The destroy method is called automatically by clients using [libmultiprocess](https://github.com/bitcoin-core/libmultiprocess), as [sv2-tp](https://github.com/stratum-mining/sv2-tp) does. In the Python tests it also happens when references are destroyed or go out of scope.

Based on:

- #35675

Commits:

- _refactor: move CTransactionRefComp to util/hasher_
- _mining: track non-mempool memory usage_: add `TxTemplateMap` to `BlockTemplateManager` to track how many templates contain any given transaction. This map is updated by the `BlockTemplate` constructor and destructor.
- _mining: add GetTemplateMemoryUsage()_ - loops over this map and sums up the memory footprint for transactions outside the mempool (includes a fuzzer)
- _ipc: add getMemoryLoad()_ expose this information to IPC clients and add test coverage

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -857,16 +861,40 @@ class ChainImpl : public Chain
+
+        TxTemplateMap& tx_refs{*Assert(m_tx_template_refs)};
+        // Don't track the dummy coinbase, because it can be modified in-place
+        // by submitSolution()
```
b9306b79b8f5667a2679236af8792bb1c36db817: in addition, we might be wiping the dummy coinbase from the template later: https://github.com/Sjors/bitcoin/pull/106

ismaelsadeeq reviewed (commented):
Concept ACK

I think it would be better if we have internal memory management for the mining interface IPC, since we hold on to the block templates.

I would suggest the following approach:

- Add memory budget for the mining interface.
- Introduce a tracking list of recently built block templates and total memory usage.
- Add templates to the list and increment the memory usage after every `createnewblock` or `waitnext` return.
- Whenever the memory budget is exhausted, we should release templates in FIFO order.

I think since we create a new template after a time interval elapses even if fees increase and that interval is usually enough for the client to receive and distribute the template to miners, this mechanism should be safe as the miners have long switch to most recent template when the budget elapsed because of the time interval being used in between returns of waitnext.

Mining interface clients should also handle their own memory internally.

Currently, I don’t see much use for the exposed `getMemoryLoad` method. In my opinion, we should not rely on the IPC client to manage our memory.

ismaelsadeeq commented:
> It seems counter intuitive, but from a memory management perspective IPC clients are treated no different than our own code. And if we started FIFO deleting templates that are used by our own code, we'd crash.

IMHO I think we should separate that, and treat clients differently from our own code, because they are different codebases and separate applications with their own memory.

> Maybe there are 100 downstream ASICs, one of which is very slow at loading templates, so it’s only given a new template when the tip changes, not when there’s a fee change. In that scenario you have a specific template that the client wants to “defend” at all costs.

I see your point but I don’t think that’s a realistic scenario, and I think we shouldn’t design software to be one-size-fits-all.
If you want to use only single block templates, then use `createnewblock` and create a new block template and mine that continuously until the chain tip changes or you mine a block.

`waitNext` returning indicates that we assume your miners are switching from the block they are currently mining to the new one they receive.
Depending on the budget (which I assume is large), many templates would need to be returned before we exhaust it.

Delegating template eviction responsibility to the client can put us in a situation where they handle it poorly and cause us to OOM (but I guess your argument is that we rather take that chance than being in a situation where we make miners potentially lose on rewards).
However I think if there is a clean separation of concerns between the Bitcoin Core node and its clients and clear interface definition and expectations that should not happen, and I believe the mining interface should not differ in that respect.
Otherwise, if we do want a one-size-fits-all solution capable of handling the scenario you described, we should rethink the design entirely and revert to an approach where we do not retain block templates.

Sjors commented:
I restructured the implementation and commits a bit.


The `TxTemplateMap` now lives on the `NodeContext` rather than `MinerImpl` (interface). This reflects the fact that we want to track the global memory footprint instead of per client. It's a lightweight member `template_tx_refs` which should be easy to fold into a block template manager later.

It's also [less code churn](https://github.com/bitcoin/bitcoin/compare/3b77529f74b76ea52c1b7b4ac4edacb0947ff202..24592b737fb80f07276a88fbcf8eaf0ee64c7e49#diff-0ef8ae12c6e9ef2accc78537f42612b3267e8a7c45dc7e9eb998f797e79f2e95) because I don't have to touch the `BlockTemplateImpl` constructor.

It also made it easier to move `GetTemplateMemoryUsage` from `interface.cpp` to `miner.cpp`, where it's more reusable.

This in turn let me split out a separate commit that introduces the actual `getMemoryLoad()` interface method. So even if we decide against including that method, the rest of the PR should be useful. However I do think it's worth keeping, it's already been a helpful debugging and monitoring tool.

I added some comments to point out that we don't hold a `mempool.cs` lock during the calculation because we don't need an accurate result (mempool drift) and we don't want to bog down transaction relay with a potentially long lock (1-3ms in my testing so far).

brunoerg commented on src/node/interfaces.cpp:
```diff
@@ -866,7 +868,22 @@ class BlockTemplateImpl : public BlockTemplate
+        for (const CTransactionRef& tx : m_block_template->block.vtx | std::views::drop(1)) {
+            auto ref_count{m_node.template_tx_refs.find(tx)};
+            if (!Assume(ref_count != m_node.template_tx_refs.end())) break;
+            if (--ref_count->second == 0) {
```
Just ran a mutation testing for this PR and this is the only unkilled mutant - feel free to ignore if it doesn't make sense to address:

```diff
diff --git a/src/node/interfaces.cpp b/src/node/interfaces.cpp
index f734296b24..13ddbb672f 100644
--- a/src/node/interfaces.cpp
+++ b/src/node/interfaces.cpp
@@ -880,7 +880,7 @@ public:
         for (const CTransactionRef& tx : m_block_template->block.vtx | std::views::drop(1)) {
             auto ref_count{m_node.template_tx_refs.find(tx)};
             if (!Assume(ref_count != m_node.template_tx_refs.end())) break;
-            if (--ref_count->second == 0) {
+            if (1==1) {
                 m_node.template_tx_refs.erase(ref_count);
             }
         }
```

brunoerg replied:
Nice, thank you!

Sjors commented:
Here's a slightly more realistic plot from last night on a well connected node running on an Intel i5-8400:

<img width="3842" height="1369" alt="getmemoryload-scatter" src="https://github.com/user-attachments/assets/f1ae2d88-73fa-4d2c-bd5f-bce5a168eeee" />

It's connected to DMND pool, declaring custom templates and getting them approved, but not actually mining. Due to their rate limiting I set `-sv2interval=20`, so if fees go up, it waits at least 20 seconds before generating a new template. It does not wait when the tip changes.

The machine also runs a lightning node and BTCPay so the moment block comes in the system is quite busy.

ryanofsky commented on src/rpc/mining.cpp:
```diff
@@ -858,7 +858,7 @@ static RPCHelpMan getblocktemplate()
     static CBlockIndex* pindexPrev;
     static int64_t time_start;
-    static std::unique_ptr<BlockTemplate> block_template;
+    std::unique_ptr<BlockTemplate>& block_template{node.gbt_template};
```
In commit "rpc: move static block_template to node context" (a5eee29fd7d177f57c78da9773cb656a129de839)

I think it would actually be nice to move all these static variables to a struct or class like @ismaelsadeeq's  BlockTemplateCache. But this could be a followup

ryanofsky commented on test/functional/interface_ipc.py:
```diff
@@ -190,23 +197,28 @@ async def async_routine():
             assert_equal(template7.to_dict(), {})
 
+            self.log.debug("Memory load should be zero because there was no mempool churn")
+            with self.nodes[0].assert_debug_log(["Calculate template transaction reference memory footprint"]):
```
In commit "ipc: add getMemoryLoad()" (e8f8f7f677bcde0179526be3ed9a657c44998b93)

Seems ok to assert this log message is logged, but I'm wondering if there was a particular reason for doing this. Was the idea to pair the LOG_TIME_MILLIS_WITH_CATEGORY and assert_debug_log calls together?

Sjors replied:
This just checks that bench logging happens.

ryanofsky reviewed (commented):
Code review e8f8f7f677bcde0179526be3ed9a657c44998b93. This looks good except for a thread safety issue I think you can address by adding a mutex.

re: https://github.com/bitcoin/bitcoin/pull/33922#issuecomment-3611468373

> > Would be good if this said templates are also released
> 
> Added a sentence to the PR description.

Sorry, I should have made a more specific suggestion. The problem is is that this sentence is not accurate: "templates are not released until the client disconnects or calls the destroy() method." Templates will be released if the client drops references to them, even if it never disconnects or calls destroy. I would just change it to "templates are not released until the client drops references to them, or calls the template destroy method, or disconnects"

Sjors commented on src/node/context.h:
```diff
@@ -67,7 +69,11 @@ struct NodeContext {
-    //! Cache latest getblocktemplate result for BIP 22 long polling
+    //! Track how many templates (which we hold on to on behalf of connected IPC
+    //! clients) are referencing each transaction.
+    TxTemplateMap template_tx_refs;
```
To limit the scope of this PR, I only added the mutex, but called it `template_state_mutex` in anticipation.

Sjors commented:
Rebased after #33819

vasild commented on src/node/context.h:
```diff
@@ -66,6 +70,13 @@ struct NodeContext {
+    TxTemplateMap template_tx_refs GUARDED_BY(template_state_mutex);
+    //! Cache latest getblocktemplate result for BIP 22 long polling. Must be cleared
+    //! before template_tx_refs.
+    std::unique_ptr<interfaces::BlockTemplate> gbt_template;
```
Is "gbt_template" supposed to mean "get block template template"?

vasild commented on src/node/miner.h:
```diff
@@ -14,6 +14,7 @@
 #include <util/feefrac.h>
```
In the commit message of ce781758df81bca7bc3b60b8e02e36064634439e `ipc: add getMemoryLoad()`

"which can later be expand to" ->
"which can later be expanded to"

vasild reviewed (approved):
ACK c37d715c1d32020fe16ddc92f7ca95f585cc6b4b

Sjors commented on src/node/types.h:
```diff
@@ -172,6 +173,11 @@ enum class TxBroadcast : uint8_t {
+/*
+ * Map how many templates refer to each transaction reference.
+ */
+using TxTemplateMap = std::map<CTransactionRef, size_t>;
```
Done both.

Assuming one template per second for two hours (long block interval), less than 10k.

vasild replied:
If this needs to be resolved, then the below should do it:

```diff
diff --git i/src/node/types.h w/src/node/types.h
index 164667772a..7bb187ac1c 100644
--- i/src/node/types.h
+++ w/src/node/types.h
@@ -11,21 +11,23 @@
 //! files.
 
 #ifndef BITCOIN_NODE_TYPES_H
 #define BITCOIN_NODE_TYPES_H
 
 #include <consensus/amount.h>
-#include <cstddef>
-#include <cstdint>
-#include <optional>
 #include <policy/policy.h>
 #include <primitives/transaction.h>
 #include <script/script.h>
-#include <unordered_map>
 #include <uint256.h>
+#include <util/hasher.h>
 #include <util/time.h>
+
+#include <cstddef>
+#include <cstdint>
+#include <optional>
+#include <unordered_map>
 #include <vector>
 
 namespace node {
 enum class TransactionError {
     OK, //!< No error
     MISSING_INPUTS,
@@ -173,11 +175,11 @@ enum class TxBroadcast : uint8_t {
     NO_MEMPOOL_PRIVATE_BROADCAST,
 };
 
 /**
  * Map how many templates refer to each transaction reference.
  */
-using TxTemplateMap = std::unordered_map<CTransactionRef, size_t>;
+using TxTemplateMap = std::unordered_map<CTransactionRef, size_t, CTransactionRefSaltedHash, CTransactionRefComp>;
 
 } // namespace node
 
 #endif // BITCOIN_NODE_TYPES_H
diff --git i/src/util/hasher.h w/src/util/hasher.h
index 02c7703391..d3b77ba72d 100644
--- i/src/util/hasher.h
+++ w/src/util/hasher.h
@@ -114,7 +114,23 @@ private:
 public:
     SaltedSipHasher();
 
     size_t operator()(const std::span<const unsigned char>& script) const;
 };
 
+struct CTransactionRefSaltedHash {
+    SaltedWtxidHasher m_wtxid_hasher;
+
+    size_t operator()(const CTransactionRef& tx) const
+    {
+        return m_wtxid_hasher(tx->GetWitnessHash());
+    }
+};
+
+struct CTransactionRefComp {
+    bool operator()(const CTransactionRef& a, const CTransactionRef& b) const
+    {
+        return a->GetWitnessHash() == b->GetWitnessHash();
+    }
+};
+
 #endif // BITCOIN_UTIL_HASHER_H
```

Using a salted hash instead of barely using the first bytes of the transaction id as a hash because somebody may craft a pile of transactions with such ids as to make `unordered_map` lookup time deteriorate from `O(1)` to `O(size)`. Not sure if that is an overkill in the current use case. In `PrivateBroadcast::m_transactions` I used the simpler:

```cpp
    struct CTransactionRefHash {
        size_t operator()(const CTransactionRef& tx) const
        {   
            return static_cast<size_t>(tx->GetWitnessHash().ToUint256().GetUint64(0));
        }
    };
```
because there 1. the transactions are originating locally (do not come from untrusted sources) and 2. the expectation is to store a small number of transactions, so even `O(size)` will not hog the machine.

Sjors commented:
Rebased after #34422 in preparation for Rust testing support in https://github.com/2140-dev/bitcoin-capnp-types.

Implemented all of @vasild's nits.

vasild reviewed (approved):
ACK 04553cd6612f08d757fb20287fe111007bfb6db1

Would be good to figure out if https://github.com/bitcoin/bitcoin/pull/33922#discussion_r2894597519 needs addressing.

enirox001 reviewed (commented):
ACK 04553cd

Went through each commit:

- [x] 6f12988 - forward declaration of `BlockTemplate` in `context.h` is necessary, `mining.h` is not in the include chain
- [x] b22e535 - destruction ordering of `gbt_result` before `template_tx_refs` looks correct. break in `~BlockTemplateImpl` makes sense, and no lock inversion between `template_state_mutex` and `cs_main`
- [x] 1d7497f - `mempool.cs` is intentionally not held for the full loop to avoid blocking relay, which is well documented
- [x] 9f4b744 - locking looks consistent with commit 3, `template_state_mutex` guards the map while `mempool.cs` is acquired per `exists()` call
- [x] 04553cd - `MAX_MONEY` is a good boundary case to have covered

Ran interface_ipc_mining.py and all tests passed successfully.

vasild reviewed (approved):
ACK 10acf818ed70d4cd2af864278a35a6b64de0bb16

Sjors commented:
The IWYU linter insists on `#include <variant>`: https://github.com/bitcoin/bitcoin/actions/runs/24406181800/job/71290035323?pr=33922

But that seems wrong, so I added a commit to override it.

Sjors commented:
Trying a rebase without cd1fce108dca9c02abf76cbeb009ed6c2cfa8cc9 first, to see if #34896 fixed the IWYU false positive.

Sjors commented:
The workaround is still needed. We can drop it here if #35073 lands.

Sjors commented on src/primitives/transaction.h:
```diff
@@ -22,6 +22,7 @@
 #include <string>
 #include <tuple>
 #include <utility>
+#include <variant>
```
It was an IWYU false positive, see comments below.

hebasto commented on src/primitives/transaction.h:
```diff
@@ -403,4 +404,12 @@ struct CMutableTransaction
+template<>
+struct std::hash<CTransactionRef> {
+    size_t operator()(const CTransactionRef&) const = delete;
+};
```
Are we sure about this parameter type? According to the [docs](https://en.cppreference.com/w/cpp/utility/hash/operator().html):
> `std::hash<Key>::operator()` ... takes a single argument key of type `Key`.

Should this be `CTransactionRef` instead of `const CTransactionRef&`?

Sjors replied:
Thanks, I'll look into this after figuring out the iwyu stuff.

maflcko replied:
> Should this be `CTransactionRef` instead of `const CTransactionRef&`?

No, I am sure the docs just refer to the logical type . C.f also e.g. libc++:

https://github.com/llvm/llvm-project/blob/25b0ab2d4f7a7a4b165b26d31dd563ef4dde4f17/libcxx/include/string#L3909-L3915

Sjors replied:
(this was moved to #35101)

maflcko commented on src/primitives/transaction.h:
```diff
@@ -22,6 +22,10 @@
+// IWYU erroneously suggests including <variant> instead of <memory>
+// for std::hash<CTransactionRef>.
+// See https://github.com/include-what-you-use/include-what-you-use/issues/2007.
+// IWYU pragma: no_include <variant>
```
NACK

This is wrong and harmful, as already explained in https://github.com/bitcoin/bitcoin/pull/35073#issuecomment-4258599969

Sjors replied:
Open to suggestions that you've actually tried.

maflcko replied:
My suggestion would be either commit 10acf818ed70d4cd2af864278a35a6b64de0bb16 (CI passed IIUC, so I think that counts as "tried"?) An alternative would be https://github.com/bitcoin/bitcoin/pull/35073#discussion_r3093983379, which I presume was tried by the author that posted it. I am happy to push that suggestion to CI, if you want me to try it as well.

Sjors replied:
> My suggestion would be either commit https://github.com/bitcoin/bitcoin/commit/10acf818ed70d4cd2af864278a35a6b64de0bb16 (CI passed IIUC, so I think that counts as "tried"?)

That's fair.

> An alternative would be https://github.com/bitcoin/bitcoin/pull/35073#discussion_r3093983379,

I'm trying that again now, this time with `namespace std` that was missing in my earlier attempt.

Sjors replied:
Unfortunately that leads to a `reference to 'hash' is ambiguous` (`std::__1::hash` vs `std::hash`). Let's see if b89942d29f1e71429070b2d1140d2792eb33bf47 fixes that.

10acf818ed70d4cd2af864278a35a6b64de0bb16 is of course a lot simpler, but also brittle because slightly different builds of IWYU will demand different includes (`<variant>` or `<string_view>`).

maflcko replied:
> [10acf81](https://github.com/bitcoin/bitcoin/commit/10acf818ed70d4cd2af864278a35a6b64de0bb16) is of course a lot simpler, but also brittle because slightly different builds of IWYU will demand different includes (`<variant>` or `<string_view>`).

Correct, but I think this is already true on current master: I think on current master a different build of IWYU will already yield a different result. I know it is not ideal, but I think the current approach is to let the CI spit out the diff when in doubt, then look if the diff has a backdoor or some nasty/wrong stuff, if not, just take the diff.


> Unfortunately that leads to a reference to 'hash' is ambiguous (std::__1::hash vs std::hash). Let's see if https://github.com/bitcoin/bitcoin/commit/b89942d29f1e71429070b2d1140d2792eb33bf47 fixes that.

Ah. Thx for testing. This looks a bit nasty to work around. Honestly, I'd just go with https://github.com/bitcoin/bitcoin/commit/10acf818ed70d4cd2af864278a35a6b64de0bb16 for now. If you want, I can try https://github.com/bitcoin/bitcoin/pull/35073#discussion_r3094156124 next week or so, to see if it is overall better, but I wouldn't spend too much time on this.

Sjors replied:
I found a nicer variant, will open a fresh PR.

maflcko replied:
That looks a bit verbose. I'd just go with https://github.com/bitcoin/bitcoin/commit/10acf818ed70d4cd2af864278a35a6b64de0bb16. If you really want, I'd say you can open a follow-up if and when this one is merged. But no strong opinion, up to you.

Sjors replied:
The problem with 10acf818ed70d4cd2af864278a35a6b64de0bb16 is that it only works on our CI. On e.g. my Ubuntu 25.10 with include-what-you-use 0.26, based clang version 22.1.1, that commit fails because it insists on `<string_view>`.

Ideally we shouldn't end up in the same situation as our linter job that can't be run outside a container.

Sjors replied:
(this code was changed, and also moved to #35101, so no longer relevant here)

Sjors commented:
Waiting for CI to pass. I'm probably going to move 92fd54b259fadce8c14b8207d40431aef60a8c8d and b89942d29f1e71429070b2d1140d2792eb33bf47 in a separate PR. Need to address this too: https://github.com/bitcoin/bitcoin/pull/33922#discussion_r3093561704

Sjors commented:
The safety commit for `CTransactionRef` and IWYU changes are now in #35101 and dropped from this PR.

vasild reviewed (approved):
ACK d37e0fc16c235ff511bd9485ef505f0c5615d9e4

<details>
<summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK d37e0fc16c235ff511bd9485ef505f0c5615d9e4
-----BEGIN PGP SIGNATURE-----

iQQzBAEBCAAdFiEE5k2NRWFNsHVF2czBVN8G9ktVy78FAmnvNGMACgkQVN8G9ktV
y79uSyAAmswvwjxbv/vFTrf0BERXNHfG0lEVrvURRQavQzajUq0Q0dSiUpBlklGY
Ec+zt5x6G9vSGkYeo7ZFyjV7uwhLMMKlOGxqmnrKUT4PWHYjiItpnsJPEE8iK7hc
M2Xpq7OGNiWfc+ZT81KIYYPo6v+dRBKBh+dHwdgJ/5X7nFSMru7OV0+W4uk4wfu1
xayuZ5A1jWqM/vqYt9jEtmEaPhyVma61CxiMCGoxOIuhsuKGx49gZb+AQi3zuyD0
GKViOOA+A9krE0Ntr8MZeoXRw/+loj+yDp49s/l9LJ6rG8C3o4CKDMjzTcGSbWp8
sc1dN0wohyMwOrBToEidkC+G4eN1+uRG4KDbkfy+gpbeNfzLj7lLDppmxJGXimKm
isCC9sq14Y0caSomDvLGFddtaGgmnq4AoZQLT49j6tleOaQArQv7xWwTgjiVMcG3
OhqCEHeuJJEDoLHaoVFG4MHSx2TrgukC4UtqRvASCayHslU0z/77RlkeKZD4B/k8
iaHOocAMl0m61r+Cw4ryZ0YfGfRKrOoFJ3PLmF/pcjj3iDE2Eq1vZ0pNq/ECzpSm
zOiLRt+3r5NsUphE9jm2AROKtQApaklNjCWidK7bUCRoA884y660m/TIZKJhcDj1
p/OPxjz36ZDdTV5YLQI+JOxgdnndfF1HQqjspZIHPsMOH6e+qkIDfyfW0apExHN+
yDoI1XoEPEIXJv4nsjSFWIUZajSOSTHjV96xrx6/ca3bbE5thix9DqLhhPsvtSHn
ykxzMhbwgUQND3Af2q16YUA9qhjUAo87Ixe99M6Ywnti52gqJ5zRZf4EgWy+A5kk
tm/YLkN3xdnLl3frNZVrBGHZwEbDMZdu8WwpHgnuGPH5DASjG4HcdJaXE0WhCNpc
mVq6YZfxJZ9NNGIpcV6pU4RoFALb981XlFfRhgtC+5Nyrjui/uPF0GrXGsiHSG9w
OzWNsR9ghyBtXyxnZ8iCI+uyX8TVxc6M/vU5PoVZGEoKJaBZqh+NclfgsZLnZqx0
5iwCLPAPJZjXsabuQ8Ya54ctHHySyk36ciqMpwWRpPPkGMW4Ya3HNMfZGSQmhiYb
FeaghFsHwtgmjbBgTLznSUCzZH+Ec7qKdk0vw+uPZveYHHIUrvbAnXuNOlbAiN1Y
XDDpPtLld+wZ89c7XUVM+5tM30fm8/M0w9y5ezDAeELVniU1mcPOnBaqgLBkv570
mJa7CEmLYOaSMzkVi3gx02sjqbEsRJlx9WAQvlETuljWT/GNggZ5hu6WCndHZ8zN
hPdARxrVv9D7p4ktAUcFRRcPalbszWwxrqp4/lnSpJbiYnNZDt5V1w2QniHBg6g/
AdTc9REfGHEi14IqXV3ysK2UMlBDmg==
=q3ZT
-----END PGP SIGNATURE-----
```

vasild's public key is [on openpgp.org](https://keys.openpgp.org/vks/v1/by-fingerprint/850E138BA9C0923881DFD55FCE712013582E2230)
</details>

Sjors commented:
Rebased after #34806 (trivial include conflict).

Sjors commented:
Rebased after #33966. Not sure if marking `template_state_mutex` as `mutable` is going in the right direction?

Sjors commented:
Going to mark this draft pending #33421. It seems to be in good shape and will provide a more natural home for memory accounting here.

ismaelsadeeq committed:
node: move mining_args to BlockTemplateManager

Pass the parsed mining args to BlockTemplateManager at construction
and expose them via BlockCreateArgs(), so the manager owns the
init-time block create options instead of NodeContext.

Sjors commented:
Rebased:
- bumped `getMemoryLoad` to `@8` in `mining.capnp` because #34644 added `submitBlock @7`
- temporarily added 51dc31db12 `test: refresh MiniWallet after node restart`, cherry-picked from #35403

ismaelsadeeq committed:
miner: move SubmitBlock into BlockTemplateManager

Move SubmitBlockStateCatcher and SubmitBlock from miner.cpp into
BlockTemplateManager as a member function. This groups block submission
with block creation in the same class. The function uses m_chainman
directly instead of taking it as a parameter.

ismaelsadeeq committed:
node: introduce BlockTemplateManager

Add BlockTemplateManager, a wrapper around
BlockAssembler::CreateNewBlock(), and store it in NodeContext.

Wire it into node init and test setup so it is reset before its
mempool/chainman dependencies, and update tests and fuzz setups that
rebuild chainman or mempool. Add a unit test that verifies a block
template can be created through the manager.

ismaelsadeeq committed:
interfaces: create block template via BlockTemplateManager

Route the Mining interface's createNewBlock() through
BlockTemplateManager::CreateNewTemplate() instead of constructing a
BlockAssembler directly. Merging the init-time defaults into unset
options now happens inside CreateNewTemplate(), so every caller gets
them applied.

ismaelsadeeq committed:
node: move tip and wait helpers into BlockTemplateManager

Move the mining tip lookup and block-template waiting helpers (GetTip,
WaitTipChanged, WaitAndCreateNewBlock, InterruptWait,
CooldownIfHeadersAhead) into BlockTemplateManager so the manager owns
the template waiting flow. The manager now takes KernelNotifications
at construction.

ismaelsadeeq committed:
test: fuzz BlockTemplateManager

Sjors commented:
Rebased after #35403.

Meanwhile #33421 was superseded by #35581, which I need to study in order to decide if I want to build on top of it or continue with the existing approach.

Sjors committed:
mining: track non-mempool memory usage

IPC clients can hold on to block templates indefinately, which has the
same impact as when the node holds a shared pointer to the
CBlockTemplate. Because each template in turn tracks CTransactionRefs,
transactions that are removed from the mempool will not have
their memory cleared.

This commit adds bookkeeping to the block template constructor and
destructor that will let us track the resulting memory footprint.

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

Sjors committed:
mining: add GetTemplateMemoryUsage()

Calculate the non-mempool memory footprint for template transaction
references.

Add bench logging to collect data on whether caching or simplified
heuristics are needed, such as not checking for mempool presence.

Check the calculation in the block template manager fuzz test.

Sjors commented:
I opened #35598 and dropped 3cb6c4779e20f183347c3c1d63d8abb9d2e4c4ee _test: cover feeThreshold = MAX_MONEY_ from this PR since it's unrelated and distracting.

---

I implemented the same functionality on top of https://github.com/bitcoin/bitcoin/pull/35581  in https://github.com/Sjors/bitcoin/pull/120. It fits well, although I still haven't studied PR 35581 itself. It also appears that I can drop 22f69a10df4afdc4c7b7d97844d8c245158f1ca0 `rpc: move static block_template to node context` in that design.

If #35581 gets enough support from reviewers, I'll switch this PR over to it. That's preferable to merging this and then having to move things to the `BlockTemplateManager` later, especially since this change isn't urgent.

Sjors committed:
refactor: move CTransactionRefComp to util/hasher

Sjors commented:
I switched over to the approach from https://github.com/Sjors/bitcoin/pull/120, so this branch is now based on ~a rebased~ https://github.com/bitcoin/bitcoin/pull/35675. Also had to rebase after #34020.

Expanded the new template manager fuzzer to cover `GetTemplateMemoryUsage()`.

ismaelsadeeq committed:
test: create templates via BlockTemplateManager

ismaelsadeeq committed:
node: remove NodeContext::mining and EnsureMining

ismaelsadeeq committed:
ci: enforce iwyu for block template manager

ismaelsadeeq committed:
rpc: build generation templates via BlockTemplateManager

generateblock is no longer safe for fuzzing: it previously threw at
EnsureMining (node.mining is never set in the fuzz setup), but with
EnsureBlockTemplateManager it now executes, mining and submitting a
real block. Submission mutates the chain state shared across fuzz
iterations and writes the block to disk.

ismaelsadeeq committed:
rpc: wait for tips via BlockTemplateManager

The waitforblock, waitforblockheight and waitfornewblock RPCs are no
longer safe for fuzzing: they previously threw at EnsureMining
(node.mining is never set in the fuzz setup), but with
EnsureBlockTemplateManager they now really wait. The tip never changes
during fuzzing and nothing interrupts the wait, so a call without a
timeout blocks forever.

ismaelsadeeq committed:
rpc: route getblocktemplate internals through node

Add EnsureBlockTemplateManager() and use it in getblocktemplate for tip
lookup and longpoll waiting, while using ChainstateManager directly for
test-chain and IBD checks.

ismaelsadeeq committed:
rpc: build getblocktemplate via BlockTemplateManager

ismaelsadeeq committed:
rpc: do not copy template data in getblocktemplate

The transactions, fees and sigops costs, and the coinbase outputs
are only read, so reference them from the template instead of copying
them out of it.

ismaelsadeeq committed:
rpc: only copy the header in getblocktemplate

The block copy exists solely to apply header adjustments (time, nonce,
version bits) that must not mutate the cached template. Now that the
transactions are read directly from the template, copy only the header
instead of the entire block.

Sjors commented:
Rebased to track the latest https://github.com/bitcoin/bitcoin/pull/35675.

Sjors committed:
ipc: add getMemoryLoad()

Allow IPC clients to inspect the amount of memory consumed by
non-mempool transactions in blocks.

Returns a MemoryLoad struct which can later be expanded to e.g.
include a limit.

Expand the interface_ipc.py test to demonstrate the behavior and
to illustrate how clients can call destroy() to reduce memory
pressure.
